File: src/main/java/org/example/eshop/service/AuditLogService.java
package org.example.eshop.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

@Service
public class AuditLogService {
    private final CorrelationIdService correlationIdService;
    private final Logger logger = LoggerFactory.getLogger("AUDIT");
    private final DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    public AuditLogService(CorrelationIdService correlationIdService) {
        this.correlationIdService = correlationIdService;
    }

    public void logAdminAction(String action, String entityType, Object entityId, String details) {
        String username = getCurrentUsername();
        String timestamp = LocalDateTime.now().format(dateFormatter);
        String correlationId = correlationIdService.getCorrelationId();
        if (correlationId == null) correlationId = "UNKNOWN";

        StringBuilder sb = new StringBuilder();
        sb.append("[AUDIT] ")
          .append("correlationId=").append(correlationId).append(' ')
          .append("timestamp=").append(timestamp).append(' ')
          .append("user=").append(username).append(' ')
          .append("action=").append(action).append(' ')
          .append("entityType=").append(entityType).append(' ');
        if (entityId != null) {
            sb.append("entityId=").append(entityId).append(' ');
        }
        if (details != null) {
            sb.append("details=").append(details);
        }
        logger.info(sb.toString());
    }

    public void logAdminLogin(String username, boolean success) {
        String timestamp = LocalDateTime.now().format(dateFormatter);
        String correlationId = correlationIdService.getCorrelationId();
        if (correlationId == null) correlationId = "UNKNOWN";
        String status = success ? "SUCCESS" : "FAILED";
        logger.info("[AUDIT] correlationId={} timestamp={} user={} action=LOGIN status={}", correlationId, timestamp, username, status);
    }

    public void logAdminLogout(String username) {
        String timestamp = LocalDateTime.now().format(dateFormatter);
        String correlationId = correlationIdService.getCorrelationId();
        if (correlationId == null) correlationId = "UNKNOWN";
        logger.info("[AUDIT] correlationId={} timestamp={} user={} action=LOGOUT", correlationId, timestamp, username);
    }

    private String getCurrentUsername() {
        try {
            var authentication = SecurityContextHolder.getContext().getAuthentication();
            return authentication != null ? authentication.getName() : "UNKNOWN";
        } catch (Exception e) {
            return "UNKNOWN";
        }
    }
}

File: src/main/kotlin/org/example/eshop/service/AuditLogService.kt
package org.example.eshop.service

import org.slf4j.LoggerFactory
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

@Service
class AuditLogService(
    private val correlationIdService: CorrelationIdService
) {
    
    private val logger = LoggerFactory.getLogger("AUDIT")
    private val dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
    
    fun logAdminAction(action: String, entityType: String, entityId: Any?, details: String? = null) {
        val username = getCurrentUsername()
        val timestamp = LocalDateTime.now().format(dateFormatter)
        val correlationId = correlationIdService.getCorrelationId() ?: "UNKNOWN"
        
        val logMessage = buildString {
            append("[AUDIT] ")
            append("correlationId=$correlationId ")
            append("timestamp=$timestamp ")
            append("user=$username ")
            append("action=$action ")
            append("entityType=$entityType ")
            if (entityId != null) {
                append("entityId=$entityId ")
            }
            if (details != null) {
                append("details=$details")
            }
        }
        
        logger.info(logMessage)
    }
    
    fun logAdminLogin(username: String, success: Boolean) {
        val timestamp = LocalDateTime.now().format(dateFormatter)
        val correlationId = correlationIdService.getCorrelationId() ?: "UNKNOWN"
        val status = if (success) "SUCCESS" else "FAILED"
        
        logger.info("[AUDIT] correlationId=$correlationId timestamp=$timestamp user=$username action=LOGIN status=$status")
    }
    
    fun logAdminLogout(username: String) {
        val timestamp = LocalDateTime.now().format(dateFormatter)
        val correlationId = correlationIdService.getCorrelationId() ?: "UNKNOWN"
        
        logger.info("[AUDIT] correlationId=$correlationId timestamp=$timestamp user=$username action=LOGOUT")
    }
    
    private fun getCurrentUsername(): String {
        return try {
            val authentication = SecurityContextHolder.getContext().authentication
            authentication?.name ?: "UNKNOWN"
        } catch (e: Exception) {
            "UNKNOWN"
        }
    }
}